pkg/planner/core/preprocess.go

```diff
@@ -449,6 +453,20 @@ func (p *preprocessor) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
 	case *ast.Constraint:
 		// Used in ALTER TABLE or CREATE TABLE
 		p.checkConstraintGrammar(node)
+	case *ast.ColumnName:
+		// When tidb_translate_softdelete_sql is enabled, the softdelete semantics are enabled for softdelete tables,
+		// and _tidb_softdelete_time column have special meaning, so user query should not reference the hidden column.
+		// The special semantics only apply to DML and SELECT statements, for which we implemented corresponding
+		// rewriting logic, so we only check these statements here.
+		// The _tidb_softdelete_time column can only appear in softdelete table, so we can just check the column name
+		// here without checking the TableInfo.
+		if p.sctx.GetSessionVars().SoftDeleteRewrite && node.Name.L == model.ExtraSoftDeleteTimeName.L &&
+			(p.stmtTp == TypeSelect || p.stmtTp == TypeInsert || p.stmtTp == TypeUpdate || p.stmtTp == TypeDelete ||
+				p.stmtTp == TypeLoadData || p.stmtTp == TypeImportInto || p.stmtTp == TypeNonTransactionalDML) {
+			p.err = plannererrors.ErrInternal.GenWithStack(
```

Comment:
Does it also affect the execute statement?

Reply:
Due to #65266, an edge case for `PREPARE/EXECUTE` is not handled correctly. I also added them to the test case.